processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java

```diff
@@ -447,14 +448,25 @@ private Assignment assignToCollection(Type targetType,
                         targetType );
                 }
                 else {
-                    // wrap the assignment in the setter method
-                    result = new SetterWrapper( result, method.getThrownTypes() );
 
+                    if ( method.isUpdateMethod() ) {
+                        // if the calling method is an update method and accesses a getter, make a local variable to
+                        // test NPE first.
+                        result = new LocalVarWrapper( result, method.getThrownTypes(), targetType );
+                    }
+                    else {
+                        // if not, asssign as new collecitin or direct
+                        result = new SetterWrapper( result, method.getThrownTypes() );
+                    }
+
+                    // TODO.. Reviewer, why is there a special collectionGetterName?
```

Comment:
> Hmm, it turns the setter method to the name of the getter. That might actually make sense, but I don't think we're checking if it actually exists. Hmm.

But we should be able to derive this in an alternative way as well, right? We have the `Type#Map<String, ExecutableElement> getPropertyReadAccessors()`